### PR TITLE
Quast: processs metaquast runs properly

### PR DESCRIPTION
**Pull request by vladsaveliev**
Fixing https://github.com/ewels/MultiQC/issues/487

In order to support outputs from metagenomic quast runs properly, we first need to determine if we are faced with a regular or a metaquast run. The basic directory structures for those are as follows:

Regular quast run (single genome), in MultiQC we want to parse `quast.tsv`:
```
quast.log
report.tsv
```
Metaquast run (multiple reference genomes), in MultiQC we want to parse `combined_reference/report.tsv`:
```
combined_reference/
   quast.log
   report.tsv
metaquast.log
runs_per_reference/
   reference1/
      quast.log
      report.tsv
   reference2/
      quast.log
      report.tsv
   ...
not_aligned/
   quast.log
   report.tsv
```

So the trick is to make MultiQC avoid treating all those per-reference subdirectories as full quast inputs. The only way I figured is to target `metaquast.log`, and then point to `dirname(join('combined_reference', 'report.tsv')` avoiding the standard file search mechanism. Only if `metaquast.log` is not found, go and search for all occurences of 'report.tsv' and process as a regular quast. I'm not sure how to distinguish between `report.tsv` and `combined_reference/report.tsv` using `find_log_files` only.

I'm also adding test data into https://github.com/ewels/MultiQC_TestData

### CODE REVIEW

**Comment by vladsaveliev** [comment]
Phil, regarding file skipping. There are many files to skip in a quast output folder, with patterns like:
`contigs_report_*.stderr`
`contigs_report_*.stdout`
`nucmer_output/*`
`*.html`
`*.pdf`
etc.
However it would be the best to just avoid parsing all subdirectories once we detected the basic report files `report.tsv` or `metaquast.log`. Do you think there is a nice way to do that?

**Comment by ewels** [comment]
Hi @vladsaveliev,

If I understand the problem correctly, is there any reason why you can't just find all Quast outputs as currently, but reject them if they contain the directory `runs_per_reference` in `f['root']`? eg:

```python
for f in self.find_log_files('quast'):
    if 'runs_per_reference' not in f['root'].split(os.path.sep):
        self.parse_quast_log(f)
```

In fact, you could just add `runs_per_reference` to `config.fn_ignore_dirs` and then those files will never make it to the `find_log_files` function. This then also solves the problem of crawling through lots of files.

Side note: if normal runs have `quast.tsv`, how did the pipeline find them before? As it's only looking for `report.tsv`?

Phil

File: setup.py [hunk 1]
```diff
@@ -52,7 +52,6 @@
         'future>0.14.0',
         'lzstring',
         'jinja2>=2.9',
-        'matplotlib',
```
Comment (ewels) [inline setup.py]: Any reason why you removed this? :)

Comment (vladsaveliev) [inline setup.py]: Oh, my bad. I had this removed in my fork because of some issues installing matplotlib. Sorry about that Phil, I'll put it back.

**Comment by vladsaveliev** [comment]
Phil, sorry for misleading - the "normal" runs had `report.tsv`, not `quast.tsv`.

**Comment by vladsaveliev** [comment]
Alright, `config.fn_ignore_dirs` sounds like what is needed. Will use that!

**Comment by vladsaveliev** [comment]
What's the correct way to append to `config.fn_ignore_dirs` in module level? Can't find any example of config manipulation in other modules. Is just appending to the list in quast.py before calling `self.find_log_files('quast')`, and restoring it right after, will work?

**Comment by ewels** [comment]
There isn't any official way to do it at module level. As it's a core module I'd just add to [`config_defaults.yaml`](https://github.com/ewels/MultiQC/blob/45642d6f93a86774ad11acee7420e82b1fc84fa3/multiqc/utils/config_defaults.yaml#L51-L52). It's a specific enough directory name that I can't see it negatively impacting other modules. And we want it to affect the main file search loop, as that will mean that all the other modules don't have to waste time searching through those files.

If you're interested, [here is an example](https://github.com/ewels/MultiQC_NGI/blob/master/multiqc_ngi/multiqc_ngi.py#L26-L67) of how I've modified the config defaults from an external MultiQC module in the past.

Phil

**Comment by vladsaveliev** [comment]
Also thinking about the case when use explicitly wants to point MultiQC to a subdirectory of a metaquast run, e.g. `runs_per_reference/reference1`. In this case we should be able to process it as a regular quast output. Or even worse, user manually names his regular quast directory `runs_per_reference` or `not_aligned`.

**Comment by ewels** [comment]
In that case, the user can just tell MultiQC to search the subdirectories. For example `multiqc data/runs_per_reference/*`. I think that should still work... (not certain).

The second case can happen, but the same can be said for a lot of other MultiQC settings. But sensible defaults should win over edge cases where possible. A lot can be mitigated by just writing about this in the documentation so that when stuff doesn't work people can find out why.

Phil

**Comment by vladsaveliev** [comment]
Yeah, I studied MultiQC_NGI project well and used the ideas in bcbio-nextgen and AZ plugins. In AZ, I even put the settings into yaml and reading them in a `config_loaded` hook, and use `config.update_dict` to override: https://github.com/vladsaveliev/MultiQC_az/blob/master/multiqc_az/multiqc_az.py#L28, but that's clearly overkill for quast. Anyway, I got the idea - will just add the directories `config_defaults.yaml`. Thanks for all clarifications!

**Comment by vladsaveliev** [comment]
Phil, I think I'm done (hopefully).

File: multiqc/modules/quast/quast.py [hunk 2]
```diff
@@ -239,30 +239,34 @@ def quast_contigs_barplot(self):
         """ Make a bar plot showing the number and length of contigs for each assembly """
 
         # Prep the data
+        import re
```
Comment (ewels) [inline multiqc/modules/quast/quast.py]: I think this is already being imported on line 8?

**Comment by ewels** [comment]
Will merge and strip the extra line myself as it's so minor..

**Comment by vladsaveliev** [comment]
I just done that too... Hope it will not cause merging clashes :)

**Comment by ewels** [comment]
Ah great, no problem.

Just testing and I ([and travis](https://travis-ci.org/ewels/MultiQC/jobs/252517780#L662-L669)) get warnings about `Not all predicted gene thresholds available for "samplename"`.  In the report I then get the error `Error - was not able to plot data.` with no plot.

Any ideas what's going on here?

Phil

**Comment by ewels** [comment]
I think the plot is failing because it's being given a bunch of empty sample dictionaries to plot. Adding a `if len(p) > 0:` check to avoid these gives a plot in the report. However, it only has a single sample (`contigs`).

**Comment by vladsaveliev** [comment]
There are several test data folders for quast, and MultiQC is running on all of them at once. Some of them were run with gene prediction turned on, some of them with it turned off. The module sees that and throws a warning, because within a single run we expect to have consistent settings.

**Comment by ewels** [comment]
Ok, so it's expected that these datasets will have missing data for this plot? In that case then just that additional check should be fine. But it's needed to avoid breaking the plot.

In fairness, I should probably make the core code able to skip past empty samples. I'll look into that.

Phil

**Comment by ewels** [comment]
Ah, now it's working. Maybe the core code does already handle empty samples then and it was something I did. Apologies, ignore me..

Phil

**Comment by vladsaveliev** [comment]
Ah, so the core works normally with empty samples? Ok!

Thanks for merging!

Also I was thinking that mentioning metaquast in docs won't hurt. Ugh, always forget about the docs. Something like that:
```
For MetaQUAST runs (`metaquast.py`), `combined_reference/report.tsv` 
is parsed, and folders `runs_per_reference` and `not_aligned` are ignored. 
If you want to run MultiQC against auxiliary MetaQUAST runs, point to 
the report.tsv explicitly, e.g.:
```
multiqc runs_per_reference/reference_1/report.tsv
```
I'll make a new PR?

**Comment by ewels** [comment]
Ah good thinking, no worries - I'll add that for you now quickly..

Phil

**Comment by ewels** [comment]
Done: http://multiqc.info/docs/#quast

**Pull request merged**